### PR TITLE
Ear7h issue 403 "Recursive environment variable support"

### PR DESCRIPTION
**Pull request by ARolek**
This is a copy of @ear7h PR which brings in "Recursive environment variable support" but with fixes to align it with several changes which were recently pushed into the v0.7.0 branch.

### CODE REVIEW

File: cache/redis/redis_test.go [hunk 1]
```diff
@@ -32,50 +55,59 @@ func TestNew(t *testing.T) {
...
 			},
-			errMatch: "connection refused",
+			expectedErr: &net.OpError{
+				Op:  "dial",
+				Net: "tcp",
+				Addr: &net.TCPAddr{
+					IP:   net.ParseIP("127.0.0.1"),
+					Port: 6000,
+				},
+				Err: syscall.ECONNREFUSED,
```
Comment (ear7h) [inline cache/redis/redis_test.go]: I changed this to use the the variable defined in the `syscall` package

File: cache/redis/redis_test.go [hunk 2]
```diff
@@ -32,50 +55,59 @@ func TestNew(t *testing.T) {
...
-				continue
-			}
-			t.Errorf("[%v] unexpected err, expected to find %q in %q", i, tc.errMatch, err)
-			continue
-		}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			fn(t, tc)
+		})
```
Comment (ear7h) [inline cache/redis/redis_test.go]: Thanks for cleaning this up

File: internal/env/map.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+package env
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-spatial/tegola/internal/dict"
+)
+
+type Map map[string]interface{}
```
Comment (ear7h) [inline internal/env/map.go]: Should be changed to `env.Dict`. `map` is inconsistent with the other similar package and I think the word "map" could become ambiguous in the presence of geographical maps.

**Comment by ear7h** [comment]
@ARolek LGTM, thank you for the fixes

**Comment by ear7h** [comment]
@ARolek Those are good points, I think it would be the most sane to implement that sort of parsing. Breaking apart the `Dicter` would make things messy when it comes to actually making use of it. For example, the `provider.Init` should't have have logic around the capabilities of a given map-like implementation. I can add this sort of parsing in a separate PR.

Also, I recall that we also need to put guards around the int <-> uint casting.

**Comment by JivanAmara** [comment]
The `Dicter` interface can't go in `internal/` since we plan to break the providers out at some point.  Should we use a dicter for `NewProvider()` calls or leave them as `map[string]interface{}`?

**Comment by ear7h** [comment]
@JivanAmara the plan is to break env and dict into a new repo also. I think this pr should go into v0.7.0 and in the next release we can take all these packages out

**Comment by gdey** [comment]
We should leave it where it is. When we break out the provider packages we can decide then what to do. I'm not sure the provider or the Dicter is in a state currently for public consumption.

File: internal/env/dict_test.go [hunk 4]
```diff
@@ -128,6 +163,31 @@ func TestDict(t *testing.T) {
...
+			dict: env.Dict{
+				"bool_slice": "${TEST_BOOL}",
+			},
+			envVars: map[string]string{
+				"TEST_BOOL": "true, false",
+			},
+			key:      "bool_slice",
+			expected: []bool{true, false},
+		},
+		"bool slice no env": {
```
Comment (ear7h) [inline internal/env/dict_test.go]: Add test case with `[]bool` in dict

File: internal/env/parse.go [hunk 5]
```diff
@@ -126,3 +201,26 @@ func ParseFloat(v interface{}) (*float64, error) {
...
 }
+
+func ParseFloatSlice(val string) ([]float64, error) {
+	// replace the env vars
+	str, err := replaceEnvVar(val)
+	if err != nil {
+		return nil, err
+	}
+
+	var floats []float64
```
Comment (ear7h) [inline internal/env/parse.go]: Preallocate slice, `Ts := make([]T, 0, len(vals))`

File: internal/env/parse.go [hunk 6]
```diff
@@ -103,6 +155,29 @@ func ParseUint(v interface{}) (*uint, error) {
...
 }
 
+func ParseUintSlice(val string) ([]uint, error) {
+	// replace the env vars
+	str, err := replaceEnvVar(val)
+	if err != nil {
+		return nil, err
+	}
+
+	var uints []uint
```
Comment (ear7h) [inline internal/env/parse.go]: Preallocate slice, `Ts := make([]T, 0, len(vals))`

File: internal/env/parse.go [hunk 7]
```diff
@@ -72,6 +107,29 @@ func ParseInt(v interface{}) (*int, error) {
...
 }
 
+func ParseIntSlice(val string) ([]int, error) {
+	// replace the env vars
+	str, err := replaceEnvVar(val)
+	if err != nil {
+		return nil, err
+	}
+
+	var ints []int
```
Comment (ear7h) [inline internal/env/parse.go]: Preallocate slice, `Ts := make([]T, 0, len(vals))`

File: internal/env/parse.go [hunk 8]
```diff
@@ -42,6 +60,29 @@ func ParseBool(v interface{}) (*bool, error) {
...
 }
 
+func ParseBoolSlice(val string) ([]bool, error) {
+	// replace the env vars
+	str, err := replaceEnvVar(val)
+	if err != nil {
+		return nil, err
+	}
+
+	var bools []bool
```
Comment (ear7h) [inline internal/env/parse.go]: Preallocate slice, `Ts := make([]T, 0, len(vals))`

File: internal/env/parse.go [hunk 9]
```diff
@@ -21,6 +22,23 @@ func ParseString(v interface{}) (*string, error) {
...
 }
 
+func ParseStringSlice(val string) ([]string, error) {
+	// replace the env vars
+	str, err := replaceEnvVar(val)
+	if err != nil {
+		return nil, err
+	}
+
+	var vals []string
```
Comment (ear7h) [inline internal/env/parse.go]: Preallocate slice, `Ts := make([]T, len(vals))`

File: internal/env/dict.go [hunk 10]
```diff
@@ -240,12 +282,26 @@ func (d Dict) Uint(key string, def *uint) (v uint, err error) {
...
+		v, err = ParseUintSlice(val.(string))
+		if err != nil {
+			switch err.(type) {
+			case ErrEnvVar:
+				return v, err
+			default:
+				return v, dict.ErrKeyType{Key: key, Value: val, T: reflect.TypeOf(v)}
+			}
+		}
+	case []int:
```
[the remark this answers is not in the document]
Comment (ARolek) [reply internal/env/dict.go]: yep, good catch.

**Pull request merged**